### ISSUE TITLE
phpunit: load modules/plugins more realistically

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/dev-lib#161]
During the PHPUnit tests we load the module manually, instead of how it would on a normal WordPress install. The same goes for how WordPoints loads itself. The issue with this is that it won't bring any load-order issues to light, or other things like that. Ideally, the module should be loaded natively, just as it would in real life. We do have to figure out how to insure that its install routine gets run though.

See https://github.com/WordPoints/buddypress/issues/12#issuecomment-244560822 for background.

**Comment by JDGrimes** [comment]
One potential drawback to this, of course, is that it requires that the plugins/modules in quesetion be actually in the plugins/modules directory. They can be symlinked there though, so it isn't such a bug deal. Although I just happened to think of https://github.com/WordPoints/wordpoints/issues/430#issuecomment-235890236, which means that the modules directory isn't set to the module directory of the test site when the tests are being run. So that is something that we really need to fix.

**Comment by JDGrimes** [comment]
> We do have to figure out how to insure that its install routine gets run though.

I had speculated that maybe there was a way to do this along with the WordPress installation code, but there really isn't, as far as I can see. We'd have to hook in after that somewhere and run it separately. Which if we did in a separate process would add some time to the set up portion of the test suite.

**Comment by JDGrimes** [comment]
I guess we'll just have to steal the code to install the plugins and module from the uninstall testers.

**Comment by JDGrimes** [comment]
Maybe this bootstrap should really be separate from this project though. I have a feeling that some other devs might adopt it if it was.

**Comment by JDGrimes** [comment]
And vòila: https://github.com/JDGrimes/wp-plugin-phpunit-bootstrap

**Comment by JDGrimes** [comment]
I've released version 0.1.0 of what is now affectionately called [WPPPB](https://github.com/JDGrimes/wpppb). I have also added a special loader to WordPoints core, that supports modules and components, as well as plugins. I'm not sure whether that should ultimately get moved to here or not. I'll let this stay open for now, until things cool off a bit.

**Comment by JDGrimes** [comment]
We do need to update our PHPUnit bootstrap for the modules to utilize the new loader when it is available.

**Comment by JDGrimes** [comment]
The issue is that we have to load the loader. We have two options there. We can load it from WordPoints, or bundle it with the dev-lib. If we bundle it with the dev-lib, then the dev-lib ends up with dependency on a particular version of WPPPB (although it might already have that anyway, but not in PHP code, in the Travis config). Anyway, it also leads to the question of whether we should just use WPPPB as installed in WordPoints, or whether the module has to bundle WPPPB itself (as a composer dep). I guess that when we run Travis builds, we don't run the composer install command for WordPoints, just the module. We could change that of course, but it would add time to the tests. Modules should have their own versions of dependencies I guess, not just use WordPoints's version. That also makes us freer to upgrade WordPoints's version without breaking modules' tests. However, that would mean that we should have the module/dev-lib load WordPoints, rather than having WordPoints load itself (that is, register itself with the loader) as it does now. Or, I guess we wouldn't have to, but it would would mean that something could break if we update the version of WPPPB that we use in WordPoints (though I guess maybe that isn't very likely).

**Comment by JDGrimes** [comment]
I guess we are really getting into the bigger question here of whether we should use the WordPoints bootstrap in our module bootstrap. We always have, but it now makes things difficult. There are good reasons to use it though, like the fact that if we need to update the WordPoints bootstrap, it is automatically updated in the dev lib also. However, it can have the opposite effect, and make it so that we can't update the bootstrap for fear of breaking modules that are using an old version of the dev lib. So I have been trying to think of a way that we can get the best of both worlds.

Really, on part of the problem is that we have to load all of the WordPoints bootstrap. Or rather, the WordPoints bootstrap doesn't just load WordPoints's test related stuff, or WordPoints itself, but it also actually loads the WordPress bootstrap. And we can't prevent that. Before it was necessary that some things be loaded after WordPress was loaded, and that other things be loaded before (to load WordPoints). But no it is not so important, because we have autoloading. So maybe we can break this down:

- Loading WordPress.
- Things that need to happen after WordPress is loaded.
- Things that need to happen before WordPress is loaded.
- Things that relate to the tests only.
- Things that relate to loading WordPoints.

So basically, maybe it would be better if we split WordPoints's bootstrap into several different files. We could keep the `bootstrap.php` itself around for WordPoints to use, but it would mostly just include other files:

- `autoload.php` to set up autoloading and maybe load anything that can't be autoloaded.
- `after-load.php` to set up things after WordPress is loaded.

**Comment by JDGrimes** [comment]
However, an issue here in regard to composer is that WordPoints now automatically includes the composer autoloader in its bootstrap. Of course, that won't exist if we don't run composer init. We could just have WordPoints only include the autoload file from composer if it exists/the module tests aren't being run. However, that would require us to include WPPPB with the module, which I guess, after all, is what we are planning at this point anyway.

**Comment by JDGrimes** [comment]
What if instead of splitting this out into different files, we split the file into functions that were hooked to different actions? Possibly some could even be methods on the loader class. That would allow us to be more granular. So then we can have a set up file that doesn't actually execute anything, it just loads and hooks up the functions. So we can then load just one file but things will be executed on the correct actions. The issue is that we can't call `do_action()` until after WordPress is loaded, which is an issue.

**Comment by JDGrimes** [comment]
Maybe what we need to do is have `add_action()` and `do_action()` method to the WordPoints loader class.

**Comment by JDGrimes** [comment]
One issue with having each module specify WPPPB as a composer dep is that it will need to also use the 5.2 autoloader package if it wants to use the composer autoloader and run against 5.2. But I guess we could create the `composer.json` file for them anyway, so it wouldn't be a big deal.

**Comment by JDGrimes** [comment]
I think that maybe we do need to make the loader a part of the dev lib instead of being a part of WordPoints. That way it would be possible for us to use it even with older versions of WordPoints. But then, I guess that really wouldn't work anyway, because WordPoints would still be loaded using the old method. But no, we decide how to load WordPoints, it doesn't load itself when the module tests are running (or that is, it did not use to). So that would be feasible, although it would require us to detect the situation and add WordPoints to the list of things to load. Because otherwise WordPoints wouldn't get loaded at all.

**Comment by JDGrimes** [comment]
So that is possible, but on second thought, I'm wondering what the real long-term benefits are of bundling the loader with the dev lib instead of with WordPoints. After 2.3 is released, compatibility with 2.1 will no longer be a concern. I guess the main issue to consider is compatibility between the version of WPPPB and the version of the loader used. If the module is supplying the WPPPB version, then it would make sense for it to also supply the WordPoints-specific loader, via whatever version of the dev lib that it bundles.

But then we still have the issue of compatibility between the version of the loader being used, and the WordPoints bootstrap. Unless we go back to having WordPoints leave loading it itself up to the dev lib. The only issue there though is if anything changes in regard to how WordPoints should be loaded. But I guess really that isn't such a big problem, since the only thing likely to change is for components to be added, and that won't happen very often. Though on the other hand, I suppose that changes in the loader aren't likely to be breaking very often either, so letting WordPoints load itself shouldn't really be a problem. Really, I'd say it is more likely for us to add new components than for the loader to have breaking changes. The former is almost certain, the latter may never happen, and isn't currently planned. Although by that same argument we might conclude that it isn't necessary to have the module supply the WordPoints loader either, since it is unlikely for it to become incompatible with WPPPB. Really, the greater concern there would be that WordPoints would expect the loader to behave differently or something, and there would be an issue between the loader and the bootstrap than that there would be an issue between the loader and WPPPB. But isn't that the opposite of what we just said? Well, I guess the difference there is breaking changes in the loader vs WordPoints wanting to use new features of the loader.

**Comment by JDGrimes** [comment]
I say leave things as they are for now, let WordPoints keep bundling the loader for a while, and we'll learn from experience what breaks.

**Comment by JDGrimes** [comment]
We are forcing modules to install a composer dependency just to be able to run the tests. But I guess we already do that for the uninstall tester. But not every module uses that. Still, like we said, we can bundle the `composer.json` and even automatically run `comoser install` for them if we wanted to.

**Comment by JDGrimes** [comment]
I guess we really don't have to be backward compatible about all of this if we don't want to. But we'd have to call this 3.0.0, of course.

**Comment by JDGrimes** [comment]
> Still, like we said, we can bundle the composer.json and even automatically run comoser install for them if we wanted to.

Instead, why not just run `composer require` and then `wpppb-init`? I guess though that if we want to also provide the 5.2 autoloader it will be easier to bundle the `composer.json` and run install. But we should run `wpppb-init`, I guess. Though some of what it will create will be plugin-specific, and not for modules. I guess maybe we don't really need much of it anyway, and we could just copy over the parts we do.

**Comment by JDGrimes** [comment]
We used to load the module automatically, but now that we are using the loader we have to register the module with the loader, rather than loading it by including the file directly. Which mean we need to know the basename path of the module symlink, not the hard folder itself. However, we're able to guess this from the module file name.

Eventually we'll make it so that the dev lib also loads the module and its dependencies automatically. In the mean time modules can just place the loader code in `functions.php`, even though that isn't ideal. (They can't be in `bootstrap.php` because that is loaded *after* WordPress.)

---

Note that as far as our own WPPPB initialization script, we'll need to remember make it so that the uninstall tests config points to the correct bootstrap.

**Comment by JDGrimes** [comment]
One more thing here, is that I guess we haven't considered what to do with the module uninstall tester. I suppose we either need to update it to integrate with WPPPB instead of the plugin uninstall tester, or we need to just merge it into this dev-lib. I'm not sure why we originally had it separate anyway, other than maybe I hadn't even created the dev-lib yet and it was simplest just to fork the plugin uninstall tester and go from there.

**Comment by JDGrimes** [comment]
One thing that we need to do is make it so that the correct message is shown about running/not running the uninstall tests. Right now both the plugin and module messages will be shown. In order to only show the module message when uninstalling modules, we need to override the `running_uninstall_tests()` method of the loader. We can do that in the WordPoints loader, but the problem is that that is bundled with WordPoints itself. So unless we bundle this stuff for module uninstall testing with WordPoints as well, we're going to have a fragile framework that is scattered through several different libraries.

However, that said, we don't have to actually bundle all of it, we can override that method alone very simply without even introducing a child `PHPUnit_GetOpt` class, the small bit of logic could just go into that method directly. That wouldn't be future-proof, but I suppose if we are worried about this changing (and that really isn't that likely), we could still utilize such a class if available, even if we don't bundle it with WordPoints.

**Comment by JDGrimes** [comment]
Maybe the best thing to do is just to bundle the entire PHPUnit bootstrap for WordPoints into the dev-lib instead. Even the parent testcases, possibly. That way everything would be available to whatever was using the dev-lib right away, instead of being tied tot he WordPoints version. There are things, of course, which are tied to the WordPoints version though, and for those we'd have to check what version of WordPoints was being tested against and act accordingly. However, that still might be better than what we're doing now, so long as it doesn't stifle innovation too much. Some of the features of the WordPoints testcases ought to graduate up to WPPPB anyway.

**Comment by JDGrimes** [comment]
For the moment though, I'll update the loader bundled with WordPoints as suggested, and bundle the module uninstall testcase with the dev-lib.

**Comment by JDGrimes** [comment]
Closing this as fixed.
